### DISCUSSION TITLE
how to start fixtures project with browser mode?

### DISCUSSION THREAD

**Discussion by HomyeeKing** [vitest-dev/vitest#5343] [Q&A]
my config like this 
```ts
import path from 'path';
import { defineConfig } from 'vitest/config';

export default defineConfig({
  resolve: {
    alias: {
      '@': path.resolve(__dirname, './src'),
    },
  },
  test: {
    globals: true,
    coverage: {
      reportsDirectory: './.ci/client',
      include: ['src/**/*.ts', '!src/{app,types,type-utils}.ts'],
      reporter: ['json', 'json-summary', 'text'],
    },
    browser: {
      enabled: true,
      name: 'chrome',
      headless: true,
    },
  },
});

```
and have `fixture/` folder contains relative test environment and want to start project before testing
```ts
//  foo.test.ts
import { execSync } from 'child_process';
import path from 'path';

describe(() => {
  const rootDir = path.join(__dirname, '../fixtures/foo');
  beforeAll(async () => {
    execSync('npm start', {
      stdio: 'inherit',
      cwd: rootDir,
    });
    await new Promise((resolve) => {
      setTimeout(resolve, 3000);
    });
  });

  test('xxx', () => {
    expect(1).toBe(1);
  });
});

```
but got error `Error: Module "child_process" has been externalized for browser compatibility. Cannot access "child_process.execSync" in client code.  See https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.`

how to solve this, and what's the best practice with browser mode?

**Comment by hi-ogawa** [comment]
Test files (like your `foo.test.ts`) and `setupFiles` https://vitest.dev/config/#setupfiles run inside the test runner/isolation (which is a browser in this case), so running `execSync` doesn't work there.

If you want to make something similar to playwright's `webServer` config https://playwright.dev/docs/api/class-testconfig#test-config-web-server, then I think you can use `globalSetup` https://vitest.dev/config/#globalsetup

But, it looks like you want to run a different command per test, so `globalSetup` might not work. I don't have a solution for such use case.

**Reply by HomyeeKing** [reply]
yes, each test file has relevant test project.

**Reply by hi-ogawa** [reply]
Could you explain use cases more in detail? The short answer is that "you cannot run `execSync` in browser mode", but depending on what actually does in `execSync`, there might be a different way. 

For example, is it a one shot command to generate assets or something? or is is something like a server which you spawn before the tests and need to kill after tests?

**Reply by HomyeeKing** [reply]
The latter. To make the test more powerful, I want to start the project with specific context, then test the request,  the dom or some global variable on the page

**Reply by sheremet-va** [reply]
You cannot use Node.js API in the browser (`child_process` is a Node.js API). There is a proposal that will allow you to connect to the server in the future via "commands": https://github.com/vitest-dev/vitest/pull/5097
